### PR TITLE
[657] catch Errno 57 on MacOS from socket.shutdown()

### PR DESCRIPTION
**Pull request by trel**
There is a race condition in BSD (of which MacOS is a decendent) where socket.shutdown() only succeeds if the socket is still open.

If the socket is closed, the call fails with
OSError: [Errno 57] Socket is not connected

Linux and Windows do not error out if the socket is closed.

### CODE REVIEW

**Comment by trel** [comment]
This was tested manually... and with this additional else statement below, I was able to see the error get caught and skipped.   This code is not included in the PR.

```diff
--- a/irods/connection.py
+++ b/irods/connection.py
@@ -333,6 +333,8 @@ class Connection:
                     # OSError: [Errno 57] Socket is not connected
                     if e.errno != errno.ENOTCONN:
                         raise
+                    else:
+                        print('============= SKIPPING - Errno 57 on socket.shutdown ============')
                 self.socket.close()
         finally:
```

File: irods/connection.py
Comment (korydraughn) [inline irods/connection.py:333]: This comment should be moved above L329 since that is the function call that triggers this exception block.

**Review by korydraughn — COMMENTED** [review]
Looks good. Just one comment.

File: irods/connection.py
Comment (trel) [reply irods/connection.py]: moved and supplemented.

File: irods/connection.py
Comment (trel) [inline irods/connection.py:333]: do we have a strong opinion about where this import line lives?   here or at the top?

Comment (korydraughn) [reply irods/connection.py:333]: Keeping it where it is helps to show how narrow its usage is and I think that's a good thing.

@d-w-moore @alanking Any thoughts?

File: irods/connection.py
Comment (korydraughn) [reply irods/connection.py]: Nice. Resolving.

File: irods/connection.py
Comment (korydraughn) [reply irods/connection.py:334]: Can't hurt to do that.

File: irods/connection.py
Comment (korydraughn) [inline irods/connection.py:336]: Two questions:
1. Should this code be wrapped in an OS check (i.e. if on MacOS)?
2. Are we potentially hiding the ENOTCONN error for other cases?

Comment (trel) [reply irods/connection.py:336]: 1. mmm, maybe, but i think question 2 is more important
2. since the point of a `.shutdown()` call is to end with a no longer connected socket... i think we're fine.

File: irods/connection.py
Comment (trel) [reply irods/connection.py:334]: will add a little more

File: irods/connection.py
Comment (trel) [reply irods/connection.py:333]: - pro for 'up top' - single truth, convention, 'performance'
- pro for 'local / in the function' - easier to 'see', and easier to refactor

don't have to be different here, happy to move it up top.

File: irods/connection.py
Comment (trel) [reply irods/connection.py]: more added.

File: irods/connection.py
Comment (trel) [reply irods/connection.py]: moved it

Comment (korydraughn) [reply irods/connection.py]: Cool cool.

File: irods/connection.py
Comment (korydraughn) [reply irods/connection.py:339]: I agree. This is fine.

Comment (korydraughn) [reply irods/connection.py:339]: A Python implementation which doesn't have this bug will never trigger this unless the users is doing something extremely weird.

Resolving.

**Review by korydraughn — APPROVED** [review]
Let's wait for one more approval before squashing.

**Review by alanking — COMMENTED** [review]
Looks good. Squash em.

**Comment by trel** [comment]
squashed to one.

**Review by alanking — APPROVED** [review]
\# it

**Comment by trel** [comment]
Thanks - #'d.

**Pull request merged**
